### PR TITLE
Lights &  GPIO Switch inverted restore mode

### PR DESCRIPTION
**Pull request by frenck**
# What does this implement/fix? 

This implements an inverted restore mode for lights & the GPIO switch platforms.

Based on Tasmota's `TOGGLE` PowerOnState functionality: Restore the state, but invert the restored state: <https://tasmota.github.io/docs/PowerOnState/#poweronstate-functionality>

Example behavior from Tasmota (user video, placed with permission).

https://user-images.githubusercontent.com/195327/119051451-76edf680-b9c3-11eb-81e3-a07350c6ef0d.mov

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1190
  
# Test Environment

- [x] ESP32
- [ ] ESP8266
- [ ] Windows
- [x] Mac OS
- [x] Linux

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
switch:
  - platform: gpio
    name: 'Sonoff Basic Relay'
    pin: GPIO12
    id: relay
    restore_mode: RESTORE_INVERTED_DEFAULT_ON

light:
  - platform: binary
    name: 'Desk Lamp'
    output: light_output
    restore_mode: RESTORE_INVERTED_DEFAULT_OFF

output:
  - id: light_output
    platform: gpio
    pin: GPIO16

```

# Explain your changes

Describe your changes here to communicate to the maintainers **why we should accept this pull request**.
Very important to fill if no issue linked

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/gpio/switch/__init__.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
     "RESTORE_DEFAULT_ON": GPIOSwitchRestoreMode.GPIO_SWITCH_RESTORE_DEFAULT_ON,
     "ALWAYS_OFF": GPIOSwitchRestoreMode.GPIO_SWITCH_ALWAYS_OFF,
     "ALWAYS_ON": GPIOSwitchRestoreMode.GPIO_SWITCH_ALWAYS_ON,
+    "RESTORE_INVERTED_DEFAULT_OFF": GPIOSwitchRestoreMode.GPIO_SWITCH_RESTORE_INVERTED_DEFAULT_OFF,
```
Comment (OttoWinter) [inline esphome/components/gpio/switch/__init__.py:16]:
```suggestion
    "RESTORE_TOGGLE_DEFAULT_OFF": GPIOSwitchRestoreMode.GPIO_SWITCH_RESTORE_TOGGLE_DEFAULT_OFF,
```

Just a minor thing, but `inverted` sounds quite cryptic, where `toggle` explains the intended effect quite well (downside is that it's a verb, not an adjective)

Comment (jesserockz) [reply esphome/components/gpio/switch/__init__.py:16]: What about `toggled` ?

Comment (frenck) [reply esphome/components/gpio/switch/__init__.py:16]: I think "toggle" is an action; that doesn't cover it.
I originally wrote that first and quickly disliked it. What does it toggle?

The fact is, it restores and inverts the value :)

Comment (frenck) [reply esphome/components/gpio/switch/__init__.py:16]: Guys got this figured out already?

Please be concrete on the naming that needs to be changed.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**